### ISSUE TITLE
Auto stop when a GPX ride has completed 

### ISSUE THREAD

**Issue by z28kid** [cagnulein/qdomyos-zwift#905]
**Is your feature request related to a problem? Please describe.**
Maybe a message or autostop could be added when you've "completed" a gpx ride

**Describe the solution you'd like**
Not that it's unclear now but I think a way to stop the users gpx ride would be a great feature. it could be a message thats displayed along with an auto stop or something. 

**Describe alternatives you've considered**
a message or auto stopping the ride so that the user is then queued to upload or save the fit file.

**Additional context**

**Comment by cagnulein** [comment]
yes sure @z28kid 
are you using gpx for bike or run?
i mean for run i need also to stop the tape and it could be dangerous

**Comment by z28kid** [comment]
I use it for biking. On the run, if the tape is the treadmill belt, I don't mean you should have it stop the belt. for riding all resistance is reduced to zero. Running shouldn't be effected if the app knows what it is that you're doing. The app should know if the user is using it on a blue tooth bike or not.

**Comment by cagnulein** [comment]
reducing the resistance to 0 could be dangerous also? maybe i can set it to the start resistance setting and show a message on qz

**Comment by MatMuski** [comment]
I use a treadmill. He can make a message with the choice to stop / loop along with the sound that after 250m the end of the route to be prepared for it. No response to the message defaults to looping.

Possibly a selection option in the program settings.

**Comment by cagnulein** [comment]
yes good idea @MatMuski

**Comment by cagnulein** [comment]
@RickNY i just checked your log and actually, the training program calls the stop for treadmill that support stop command https://github.com/cagnulein/qdomyos-zwift/blob/fd32d29c87cbfe1c363fd0c6ddff920fefb65639/src/trainprogram.cpp#L674

so as @MatMuski suggested a message box or something would be required in order to leave the decision to the user. What do you think?

**Comment by RickNY** [comment]
> @RickNY i just checked your log and actually, the training program calls the stop for treadmill that support stop command
> 
> https://github.com/cagnulein/qdomyos-zwift/blob/fd32d29c87cbfe1c363fd0c6ddff920fefb65639/src/trainprogram.cpp#L674
> 
> so as @MatMuski suggested a message box or something would be required in order to leave the decision to the user. What do you think?

I guess the GPX and Peloton routines both use this same routine?  I’m just trying to determine two things — first, if the behavior of the Peloton workout just stopping is expected and second, if the user wants the flexibility of continuing the tread after the Peloton workout has ended, how to do that? I’m trying to follow the steps that @janabouc uses, but that isn’t working for me with this hard stop.

**Comment by cagnulein** [comment]
> I guess the GPX and Peloton routines both use this same routine?  

yes

> I’m just trying to determine two things — first, if the behavior of the Peloton workout just stopping is expected 

yes it's expected, i mean i just put a line that stop the treadmill at the end of the gpx or peloton workout 

> and second, if the user wants the flexibility of continuing the tread after the Peloton workout has ended, how to do that? I’m trying to follow the steps that @janabouc uses, but that isn’t working for me with this hard stop. 

@janabouc didn't have the beta. the stop command was implemented only recently. now she has also the same behavior for sure as she can confirm 

but i mean it's a single line of code, if everyone wants to don't stop the treadmill i can just remove it :)

**Comment by RickNY** [comment]
> 
> but i mean it's a single line of code, if everyone wants to don't stop the treadmill i can just remove it :)

I mean I can see where it would be useful -- but I can also see just a simple user option would be good -- either a) set a Peloton option "End workout when stop callout received" (Y/N)  or b) do the popup @MatMuski  suggested asking the user what they want to do

I guess I'm confused because in #973 you specifically said "nope, for safety reason i'm not stopping the tape" when I asked if it was normal that a GPX loops around.  And on the Horizon -- once that stop command goes out, you have to start all over to get going again with a 2nd new workout when all you want to do is just keep walking, running, etc for 10-15 more minutes

**Comment by cagnulein** [comment]
> I mean I can see where it would be useful -- but I can also see just a simple user option would be good -- either a) set a Peloton option "End workout when stop callout received" (Y/N) or b) do the popup @MatMuski suggested asking the user what they want to do

yes I agree with you. For me the simplest way to do this it's with a settings. So I'm voting for this one :)

> I guess I'm confused because in #973 you specifically said "nope, for safety reason i'm not stopping the tape" when I asked if it was normal that a GPX loops around. 

Yes, I got confused too...I didn't remember about this line, QZ started to be a very large piece of code, and it's hard to remember everything :D 

> And on the Horizon -- once that stop command goes out, you have to start all over to get going again with a 2nd new workout when all you want to do is just keep walking, running, etc for 10-15 more minutes

hah right, this complicates a lot more the thing.
Let me add the setting now!

**Comment by cagnulein** [comment]
done, it will be in the next beta version

**Comment by cagnulein** [comment]
@RickNY fixed this too! Thanks for the feedback!

**Comment by cagnulein** [comment]
@janabouc yes probably it's better to don't follow it

**Comment by RickNY** [comment]
> I'm glad you're fixing these two things that i fortunately haven't encountered because I've been off the treadmill for over a month. I planned to get back to it today.
> 
> So until the new beta goes out i should choose to NOT follow the class or i will get a disconnect when the class starts?

@janabouc - I did a quick 30 min Peloton class this morning and did set the option under Training Program options to "Stop Treadmill at the End" -- it did indeed pause the treadmill.  I did the closing stretches but then decided to see what QZ would do if I hit the start button again (since Roberto & I were testing a while to get the Horizon pause issue straightened out) -- ressuming with "Start" in QZ did indeed resume.  I do have "Continuous movement" also on, so maybe thats why it didnt disconnect?  During the "Pause" elapsed time stopped counting, but resumed once I restarted.
